syne_tune/optimizer/schedulers/botorch/botorch_searcher.py

```diff
@@ -0,0 +1,235 @@
+        assert num_init_random_draws >= 2
+        assert mode in ['min', 'max']
+        self.hp_ranges = make_hyperparameter_ranges(config_space=config_space)
+        self.mode = mode
+        self.metric_name = metric
+        self.num_minimum_observations = num_init_random_draws
+        self.points_to_evaluate = points_to_evaluate
+        self.config_seen = set()
+        self.hp_ranges = make_hyperparameter_ranges(config_space=config_space)
+        self.fantasising = fantasising
+        self.max_num_observations = max_num_observations
+        self.input_warping = input_warping
+        self.trial_configs = {}
+        self.pending_trials = set()
+        self.trial_observations = {}
+
+    def _update(self, trial_id: str, config: Dict, result: Dict):
+        trial_id = int(trial_id)
+        self.trial_observations[trial_id] = result[self.metric_name]
+        self.pending_trials.remove(trial_id)
+
+    def clone_from_state(self, state):
+        pass
+
+    def num_suggestions(self):
+        return len(self.trial_configs)
+
+    def get_config(self, trial_id: str, **kwargs):
+        trial_id = int(trial_id)
+        config_suggested = self._next_initial_config()
+
+        if config_suggested is None:
+            if self.points_to_evaluate is not None and self.num_suggestions() < len(self.points_to_evaluate):
+                # if we are not done yet with points_to_evaluate, we pick the next one from this list
+                config_suggested = self.points_to_evaluate[self.num_suggestions()]
+            else:
+                not_enough_suggestion = len(self.objectives()) < self.num_minimum_observations
+                if not_enough_suggestion:
+                    config_suggested = self._sample_random()
+                else:
+                    config_suggested = self._sample_next_candidate()
+
+        self.pending_trials.add(trial_id)
+        self.trial_configs[trial_id] = config_suggested
+        self.config_seen.add(tuple(config_suggested.values()))
+
+        return config_suggested
+
+    def _sample_next_candidate(self) -> Dict:
+        """
+        :return: a next candidate to evaluate, if possible it is obtained by fitting a GP on past data and maximizing EI
+        if this fails because of numerical difficulties with non PSD matrices, then the candidate is sampled at random.
+        """
+        try:
+            X = np.array(self._config_to_feature_matrix(self._configs_with_results()))
+            y = self.objectives()
+
+            if self.max_num_observations is not None and len(X) >= self.max_num_observations:
+                perm = self.random_state.permutation(len(X))
+                X = X[perm]
+                y = y[perm]
+                subsample = True
+            else:
+                subsample = False
+
+            X_tensor = torch.Tensor(X)
+            Y_tensor = standardize(torch.Tensor(y).reshape(-1, 1))
+            gp = self._make_gp(X_tensor=X_tensor, Y_tensor=Y_tensor)
```

Comment:
Wouldn't it be more efficient if we fit the GP only if new data is observed, instead of fitting it every time we query a new configuration?

Reply:
I think it boils down to the same since when a result is observed, then a configuration is sampled.